packages/@expo/config/src/__tests__/Config-test.ts

```diff
@@ -306,3 +306,120 @@ describe('hasUnusedStaticConfig', () => {
     expect(hasUnusedStaticConfig).toBe(false);
   });
 });
+
+describe(modifyConfigAsync, () => {
+  function createProject(
+    projectRoot: Parameters<typeof vol.fromJSON>[1],
+    files: Parameters<typeof vol.fromJSON>[0] = {}
+  ) {
+    const packageFile = {
+      name: 'testing123',
+      version: '0.1.0',
+      description: 'fake description',
+      main: 'index.js',
+    };
+
+    const expoPackageFile = {
+      name: 'expo',
+      version: '650.x.x',
+    };
+
+    vol.fromJSON(
+      {
+        'package.json': JSON.stringify(packageFile),
+        'node_modules/expo/package.json': JSON.stringify(expoPackageFile),
+        ...files,
+      },
+      projectRoot
+    );
+  }
+
+  const appFile = {
+    name: 'testing 123',
+    version: '0.1.0',
+    slug: 'testing-123',
+    sdkVersion: '100.0.0',
+  };
+
+  afterEach(() => vol.reset());
+
+  it('succeeds without existing config', async () => {
+    createProject('/no-config');
+
+    await expect(modifyConfigAsync('/no-config', { name: 'new name' })).resolves.toMatchObject({
+      type: 'success',
+      config: { name: 'new name' },
+    });
+  });
+
+  it('succeeds when modifying static config only', async () => {
+    createProject('/static-only', {
+      'app.json': JSON.stringify(appFile, null, 2),
+    });
+
+    await expect(modifyConfigAsync('/static-only', { name: 'new name' })).resolves.toMatchObject({
+      type: 'success',
+      config: { ...appFile, name: 'new name' },
+    });
+  });
+
+  it('warns when modifying dynamic config only', async () => {
+    createProject('/dynamic-only', {
+      'app.config.js': `module.exports = () => JSON.parse(\`${JSON.stringify({ ...appFile, version: '9.9.9' })}\`);`,
+    });
+
+    await expect(modifyConfigAsync('/dynamic-only', { name: 'new name' })).resolves.toMatchObject({
+      type: 'warn',
+      config: null,
+      message: expect.stringMatching(/dynamic config/),
+    });
+  });
+
+  it('warns when modifying static with object-like dynamic config', async () => {
+    createProject('/static-dynamic-object', {
+      'app.json': JSON.stringify(appFile),
+      'app.config.js': `module.exports = JSON.parse(\`${JSON.stringify({ ...appFile, version: '9.9.9' })}\`);`,
+    });
+
+    await expect(
+      modifyConfigAsync('/static-dynamic-object', { name: 'new name' })
+    ).resolves.toMatchObject({
+      type: 'warn',
+      config: null,
+      message: expect.stringMatching(/dynamic config/),
+    });
```

Comment:
Missing an assertion that the app.json was rolled back.

Reply:
Done!